app/api/controllers/RequestHandler.scala

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api.controllers
+
+import api.controllers.requestParsers.RequestParser
+import api.hateoas.{HateoasFactory, HateoasLinksFactory}
+import api.models.errors.{DownstreamError, ErrorWrapper}
+import api.models.hateoas.{HateoasData, HateoasWrapper}
+import api.models.outcomes.ResponseWrapper
+import api.models.request.RawData
+import cats.data.EitherT
+import cats.implicits._
+import play.api.http.Status
+import play.api.libs.json.{JsValue, Json, Writes}
+import play.api.mvc.Result
+import play.api.mvc.Results.InternalServerError
+import utils.Logging
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait RequestHandler[InputRaw <: RawData] {
+
+  def handleRequest(rawData: InputRaw)(implicit ctx: RequestContext, request: UserRequest[_], ec: ExecutionContext): Future[Result]
+
+}
+
+object RequestHandler {
+
+  def withParser[InputRaw <: RawData, Input](parser: RequestParser[InputRaw, Input]): ParserOnlyBuilder[InputRaw, Input] =
+    new ParserOnlyBuilder[InputRaw, Input](parser)
+
+  // Intermediate class so that the compiler can separately capture the InputRaw and Input types here, and the Output type later
+  class ParserOnlyBuilder[InputRaw <: RawData, Input] private[RequestHandler] (parser: RequestParser[InputRaw, Input]) {
+
+    def withService[Output](
+        serviceFunction: Input => Future[Either[ErrorWrapper, ResponseWrapper[Output]]]): RequestHandlerBuilder[InputRaw, Input, Output] =
```

Comment:
Kind of feels like this ought to be a service trait rather than a function "prototype". I wonder if it's time to "declare" that all services are single-method-only? (Given that they all basically are, pretty much)